File: src/main/java/org/cubexmc/railway/physics/TrainCartsBridge.java
package org.cubexmc.railway.physics;

import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodType;
import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import org.bukkit.Bukkit;
import org.bukkit.entity.Minecart;
import org.bukkit.plugin.Plugin;

/**
 * Reflection based TrainCarts integration. Only used when the TrainCarts plugin is present.
 */
final class TrainCartsBridge {

    private static final String[] PLUGIN_NAMES = {"Train_Carts", "TrainCarts"};

    static TrainCartsBridge createIfAvailable() {
        Plugin plugin = findPlugin();
        if (plugin == null) {
            return null;
        }
        try {
            return new TrainCartsBridge(plugin);
        } catch (ReflectiveOperationException ex) {
            plugin.getLogger().warning("Railway: TrainCarts integration failed: " + ex.getMessage());
            return null;
        }
    }

    private static Plugin findPlugin() {
        for (String name : PLUGIN_NAMES) {
            Plugin plugin = Bukkit.getPluginManager().getPlugin(name);
            if (plugin != null && plugin.isEnabled()) {
                return plugin;
            }
        }
        return null;
    }

    private final Plugin trainCartsPlugin;
    private final Class<?> minecartMemberStoreClass;
    private final Class<?> minecartMemberClass;
    private final Class<?> minecartGroupStoreClass;
    private final Class<?> minecartGroupClass;
    private final Class<?> trainPropertiesClass;
    private final Class<?> slowdownModeClass;

    private final MethodHandle convertHandle;
    private final MethodHandle getFromEntityHandle;
    private final MethodHandle clearGroupHandle;

    private final MethodHandle createGroupHandle;
    private final MethodHandle setForwardForceHandle;
    private final MethodHandle stopGroupHandle;
    private final MethodHandle getPropertiesHandle;
    private final MethodHandle setSpeedLimitHandle;
    private final MethodHandle setSlowingDownHandle;
    private final MethodHandle setSlowdownModeHandle;

    private final Object slowdownDisabledEnum;

    private TrainCartsBridge(Plugin plugin) throws ReflectiveOperationException {
        this.trainCartsPlugin = plugin;
        MethodHandles.Lookup lookup = MethodHandles.publicLookup();

        ClassLoader loader = plugin.getClass().getClassLoader();
        Class<?> trainCartsClass = loader.loadClass("com.bergerkiller.bukkit.tc.TrainCarts");
        this.minecartMemberStoreClass = loader.loadClass("com.bergerkiller.bukkit.tc.controller.MinecartMemberStore");
        this.minecartMemberClass = loader.loadClass("com.bergerkiller.bukkit.tc.controller.MinecartMember");
        this.minecartGroupStoreClass = loader.loadClass("com.bergerkiller.bukkit.tc.controller.MinecartGroupStore");
        this.minecartGroupClass = loader.loadClass("com.bergerkiller.bukkit.tc.controller.MinecartGroup");
        this.trainPropertiesClass = loader.loadClass("com.bergerkiller.bukkit.tc.properties.TrainProperties");
        this.slowdownModeClass = loader.loadClass("com.bergerkiller.bukkit.tc.properties.standard.type.SlowdownMode");

        this.convertHandle = lookup.findStatic(
                minecartMemberStoreClass,
                "convert",
                MethodType.methodType(minecartMemberClass, trainCartsClass, Minecart.class));

        this.getFromEntityHandle = lookup.findStatic(
                minecartMemberStoreClass,
                "getFromEntity",
                MethodType.methodType(minecartMemberClass, org.bukkit.entity.Entity.class));

        this.clearGroupHandle = lookup.findVirtual(
                minecartMemberClass,
                "clearGroup",
                MethodType.methodType(void.class));

        this.createGroupHandle = lookup.findStatic(
                minecartGroupStoreClass,
                "create",
                MethodType.methodType(minecartGroupClass, String.class, Array.newInstance(minecartMemberClass, 0).getClass()));

        this.setForwardForceHandle = lookup.findVirtual(
                minecartGroupClass,
                "setForwardForce",
                MethodType.methodType(void.class, double.class));

        this.stopGroupHandle = lookup.findVirtual(
                minecartGroupClass,
                "stop",
                MethodType.methodType(void.class));

        this.getPropertiesHandle = lookup.findVirtual(
                minecartGroupClass,
                "getProperties",
                MethodType.methodType(trainPropertiesClass));

        this.setSpeedLimitHandle = lookup.findVirtual(
                trainPropertiesClass,
                "setSpeedLimit",
                MethodType.methodType(void.class, double.class));

        this.setSlowingDownHandle = lookup.findVirtual(
                trainPropertiesClass,
                "setSlowingDown",
                MethodType.methodType(void.class, boolean.class));

        this.setSlowdownModeHandle = lookup.findVirtual(
                trainPropertiesClass,
                "setSlowingDown",
                MethodType.methodType(void.class, slowdownModeClass, boolean.class));

        Object disabledValue = null;
        Object[] slowdownConstants = slowdownModeClass.getEnumConstants();
        if (slowdownConstants != null) {
            for (Object constant : slowdownConstants) {
                if (constant != null && "DISABLED".equals(constant.toString())) {
                    disabledValue = constant;
                    break;
                }
            }
        }
        this.slowdownDisabledEnum = disabledValue;
    }

    TrainHandle attach(List<Minecart> carts, double speedLimit) {
        if (carts.isEmpty()) {
            return null;
        }
        List<Object> members = new ArrayList<>(carts.size());
        for (Minecart cart : carts) {
            if (cart == null || cart.isDead()) {
                continue;
            }
            Object member = getMember(cart);
            if (member == null) {
                member = convert(cart);
            }
            if (member == null) {
                return null;
            }
            try {
                clearGroupHandle.invoke(member);
            } catch (Throwable t) {
                return null;
            }
            members.add(member);
        }
        if (members.isEmpty()) {
            return null;
        }
        try {
            Object memberArray = Array.newInstance(minecartMemberClass, members.size());
            for (int i = 0; i < members.size(); i++) {
                Array.set(memberArray, i, members.get(i));
            }
            Object group = createGroupHandle.invokeWithArguments((Object) null, memberArray);
            if (group == null) {
                return null;
            }
            configureGroup(group, speedLimit);
            return new TrainHandle(group);
        } catch (Throwable t) {
            return null;
        }
    }

    void setSpeed(Object group, double speed) {
        try {
            setForwardForceHandle.invoke(group, speed);
        } catch (Throwable ignored) {
        }
    }

    void stop(Object group) {
        try {
            stopGroupHandle.invoke(group);
        } catch (Throwable ignored) {
        }
    }

    void cleanup(Object group) {
        if (group == null) {
            return;
        }
        try {
            stopGroupHandle.invoke(group);
        } catch (Throwable ignored) {
        }
    }

    void updateSpeedLimit(Object group, double speedLimit) {
        try {
            Object properties = getPropertiesHandle.invoke(group);
            if (properties != null) {
                setSpeedLimitHandle.invoke(properties, speedLimit);
            }
        } catch (Throwable ignored) {
        }
    }

    private Object getMember(Minecart cart) {
        try {
            Object member = getFromEntityHandle.invoke(cart);
            if (member != null) {
                return member;
            }
        } catch (Throwable ignored) {
        }
        return null;
    }

    private Object convert(Minecart cart) {
        try {
            return convertHandle.invoke(trainCartsPlugin, cart);
        } catch (Throwable t) {
            return null;
        }
    }

    private void configureGroup(Object group, double speedLimit) {
        try {
            Object properties = getPropertiesHandle.invoke(group);
            if (properties != null) {
                setSpeedLimitHandle.invoke(properties, speedLimit);
                setSlowingDownHandle.invoke(properties, false);
                if (slowdownDisabledEnum != null) {
                    setSlowdownModeHandle.invoke(properties, slowdownDisabledEnum, false);
                }
            }
        } catch (Throwable ignored) {
        }
    }

    static final class TrainHandle {
        private final Object group;
        TrainHandle(Object group) {
            this.group = Objects.requireNonNull(group);
        }

        void setSpeed(TrainCartsBridge bridge, double speed) {
            bridge.setSpeed(group, speed);
        }

        void stop(TrainCartsBridge bridge) {
            bridge.stop(group);
        }

        void cleanup(TrainCartsBridge bridge) {
            bridge.cleanup(group);
        }

        void updateSpeedLimit(TrainCartsBridge bridge, double speedLimit) {
            bridge.updateSpeedLimit(group, speedLimit);
        }
    }
}

File: src/main/java/org/cubexmc/metro/physics/TrainCartsBridge.kt
package org.cubexmc.metro.physics

import java.lang.invoke.MethodHandle
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType
import java.lang.reflect.Array as ReflectArray
import java.util.Objects
import org.bukkit.Bukkit
import org.bukkit.entity.Entity
import org.bukkit.entity.Minecart
import org.bukkit.plugin.Plugin

/** Reflection-based TrainCarts integration. Only used when the TrainCarts plugin is present. */
internal class TrainCartsBridge private constructor(plugin: Plugin) {

    private val trainCartsPlugin: Plugin = plugin
    private val minecartMemberStoreClass: Class<*>
    private val minecartMemberClass: Class<*>
    private val minecartGroupStoreClass: Class<*>
    private val minecartGroupClass: Class<*>
    private val trainPropertiesClass: Class<*>
    private val slowdownModeClass: Class<*>

    private val convertHandle: MethodHandle
    private val getFromEntityHandle: MethodHandle
    private val clearGroupHandle: MethodHandle

    private val createGroupHandle: MethodHandle
    private val setForwardForceHandle: MethodHandle
    private val stopGroupHandle: MethodHandle
    private val getPropertiesHandle: MethodHandle
    private val setSpeedLimitHandle: MethodHandle
    private val setSlowingDownHandle: MethodHandle
    private val setSlowdownModeHandle: MethodHandle

    private val slowdownDisabledEnum: Any?

    init {
        val lookup = MethodHandles.publicLookup()

        val loader = plugin.javaClass.classLoader
        val trainCartsClass = loader.loadClass("com.bergerkiller.bukkit.tc.TrainCarts")
        minecartMemberStoreClass = loader.loadClass("com.bergerkiller.bukkit.tc.controller.MinecartMemberStore")
        minecartMemberClass = loader.loadClass("com.bergerkiller.bukkit.tc.controller.MinecartMember")
        minecartGroupStoreClass = loader.loadClass("com.bergerkiller.bukkit.tc.controller.MinecartGroupStore")
        minecartGroupClass = loader.loadClass("com.bergerkiller.bukkit.tc.controller.MinecartGroup")
        trainPropertiesClass = loader.loadClass("com.bergerkiller.bukkit.tc.properties.TrainProperties")
        slowdownModeClass = loader.loadClass("com.bergerkiller.bukkit.tc.properties.standard.type.SlowdownMode")

        convertHandle = lookup.findStatic(
            minecartMemberStoreClass,
            "convert",
            MethodType.methodType(minecartMemberClass, trainCartsClass, Minecart::class.java),
        )

        getFromEntityHandle = lookup.findStatic(
            minecartMemberStoreClass,
            "getFromEntity",
            MethodType.methodType(minecartMemberClass, Entity::class.java),
        )

        clearGroupHandle = lookup.findVirtual(
            minecartMemberClass,
            "clearGroup",
            MethodType.methodType(Void.TYPE),
        )

        createGroupHandle = lookup.findStatic(
            minecartGroupStoreClass,
            "create",
            MethodType.methodType(
                minecartGroupClass,
                String::class.java,
                ReflectArray.newInstance(minecartMemberClass, 0).javaClass,
            ),
        )

        setForwardForceHandle = lookup.findVirtual(
            minecartGroupClass,
            "setForwardForce",
            MethodType.methodType(Void.TYPE, java.lang.Double.TYPE),
        )

        stopGroupHandle = lookup.findVirtual(
            minecartGroupClass,
            "stop",
            MethodType.methodType(Void.TYPE),
        )

        getPropertiesHandle = lookup.findVirtual(
            minecartGroupClass,
            "getProperties",
            MethodType.methodType(trainPropertiesClass),
        )

        setSpeedLimitHandle = lookup.findVirtual(
            trainPropertiesClass,
            "setSpeedLimit",
            MethodType.methodType(Void.TYPE, java.lang.Double.TYPE),
        )

        setSlowingDownHandle = lookup.findVirtual(
            trainPropertiesClass,
            "setSlowingDown",
            MethodType.methodType(Void.TYPE, java.lang.Boolean.TYPE),
        )

        setSlowdownModeHandle = lookup.findVirtual(
            trainPropertiesClass,
            "setSlowingDown",
            MethodType.methodType(Void.TYPE, slowdownModeClass, java.lang.Boolean.TYPE),
        )

        var disabledValue: Any? = null
        val slowdownConstants = slowdownModeClass.enumConstants
        if (slowdownConstants != null) {
            for (constant in slowdownConstants) {
                if (constant != null && constant.toString() == "DISABLED") {
                    disabledValue = constant
                    break
                }
            }
        }
        slowdownDisabledEnum = disabledValue
    }

    fun attach(carts: List<Minecart?>, speedLimit: Double): TrainHandle? {
        if (carts.isEmpty()) {
            return null
        }
        val members = ArrayList<Any>(carts.size)
        for (cart in carts) {
            if (cart == null || cart.isDead) {
                continue
            }
            var member = getMember(cart)
            if (member == null) {
                member = convert(cart)
            }
            if (member == null) {
                return null
            }
            try {
                clearGroupHandle.invoke(member)
            } catch (_: Throwable) {
                return null
            }
            members.add(member)
        }
        if (members.isEmpty()) {
            return null
        }
        return try {
            val memberArray = ReflectArray.newInstance(minecartMemberClass, members.size)
            for (index in members.indices) {
                ReflectArray.set(memberArray, index, members[index])
            }
            val group: Any? = createGroupHandle.invokeWithArguments(null as Any?, memberArray)
            if (group == null) {
                null
            } else {
                configureGroup(group, speedLimit)
                TrainHandle(group)
            }
        } catch (_: Throwable) {
            null
        }
    }

    fun setSpeed(group: Any, speed: Double) {
        try {
            setForwardForceHandle.invoke(group, speed)
        } catch (_: Throwable) {
            // Optional integration failures must not affect Railway operation.
        }
    }

    fun stop(group: Any) {
        try {
            stopGroupHandle.invoke(group)
        } catch (_: Throwable) {
            // Optional integration failures must not affect Railway operation.
        }
    }

    fun cleanup(group: Any?) {
        if (group == null) {
            return
        }
        try {
            stopGroupHandle.invoke(group)
        } catch (_: Throwable) {
            // Optional integration failures must not affect Railway operation.
        }
    }

    fun updateSpeedLimit(group: Any, speedLimit: Double) {
        try {
            val properties: Any? = getPropertiesHandle.invoke(group)
            if (properties != null) {
                setSpeedLimitHandle.invoke(properties, speedLimit)
            }
        } catch (_: Throwable) {
            // Optional integration failures must not affect Railway operation.
        }
    }

    private fun getMember(cart: Minecart): Any? {
        try {
            val member: Any? = getFromEntityHandle.invoke(cart)
            if (member != null) {
                return member
            }
        } catch (_: Throwable) {
            // Try conversion below.
        }
        return null
    }

    private fun convert(cart: Minecart): Any? =
        try {
            convertHandle.invoke(trainCartsPlugin, cart)
        } catch (_: Throwable) {
            null
        }

    private fun configureGroup(group: Any, speedLimit: Double) {
        try {
            val properties: Any? = getPropertiesHandle.invoke(group)
            if (properties != null) {
                setSpeedLimitHandle.invoke(properties, speedLimit)
                setSlowingDownHandle.invoke(properties, false)
                val disabled = slowdownDisabledEnum
                if (disabled != null) {
                    setSlowdownModeHandle.invoke(properties, disabled, false)
                }
            }
        } catch (_: Throwable) {
            // Optional integration failures must not affect Railway operation.
        }
    }

    internal class TrainHandle(group: Any) {
        private val group: Any = Objects.requireNonNull(group)

        fun setSpeed(bridge: TrainCartsBridge, speed: Double) {
            bridge.setSpeed(group, speed)
        }

        fun stop(bridge: TrainCartsBridge) {
            bridge.stop(group)
        }

        fun cleanup(bridge: TrainCartsBridge) {
            bridge.cleanup(group)
        }

        fun updateSpeedLimit(bridge: TrainCartsBridge, speedLimit: Double) {
            bridge.updateSpeedLimit(group, speedLimit)
        }
    }

    companion object {
        private val PLUGIN_NAMES = arrayOf("Train_Carts", "TrainCarts")

        @JvmStatic
        fun createIfAvailable(): TrainCartsBridge? {
            val plugin = findPlugin() ?: return null
            return try {
                TrainCartsBridge(plugin)
            } catch (exception: ReflectiveOperationException) {
                plugin.logger.warning("Metro: TrainCarts integration failed: ${exception.message}")
                null
            }
        }

        private fun findPlugin(): Plugin? {
            for (name in PLUGIN_NAMES) {
                val plugin = Bukkit.getPluginManager().getPlugin(name)
                if (plugin != null && plugin.isEnabled) {
                    return plugin
                }
            }
            return null
        }
    }
}
